### PR TITLE
Unified Alerting: Validate PostableSilence API.

### PR DESCRIPTION
**Pull request by joeblubaugh**
**What this PR does / why we need it**:
Invalid PostableSilences could be passed to the Alerting API - if they
are passed all the way down into the alertmanager data layer, they can
cause a panic. This change adds validation to avoid a panic in the
alertmanager.

**Special notes for your reviewer**:
See the "questions for reviewers" comments about validating PostableSilences with a format registry.

### CODE REVIEW

File: pkg/services/ngalert/api/api_alertmanager.go [hunk 1]
```diff
@@ -129,6 +130,19 @@ func (srv AlertmanagerSrv) RouteCreateSilence(c *models.ReqContext, postableSile
...
 	}
 
+	// Validate our silence to avoid panics
+	//
+	// Question for reviewers: do we care about the format registry? My reading
+	// of the code is that an empty format registry means "all data validates,
+	// but required fields must be present." Since we didn't evaluate before,
+	// that's the existing behavior. If we want an empty registry, I think we can
+	// just create it on each API call - it's not a very high-throughput API.
+	err := postableSilence.Validate(strfmt.NewFormats())
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/api/api_alertmanager.go:140]: I tried to look at the Alertmanager API to see what is expected there and its quite hard to follow the code. It could be something like `strfmt.Default`?

Comment (joeblubaugh) [reply pkg/services/ngalert/api/api_alertmanager.go]: I think that's equivalent to what I have here, without creating an object each time. Done.

File: pkg/services/ngalert/api/api_alertmanager.go [hunk 2]
```diff
@@ -125,6 +126,16 @@ func (srv AlertmanagerSrv) RouteGetAMStatus(c *models.ReqContext) response.Respo
 }
 
 func (srv AlertmanagerSrv) RouteCreateSilence(c *models.ReqContext, postableSilence apimodels.PostableSilence) response.Response {
+	if !c.HasUserRole(models.ROLE_EDITOR) {
+		return ErrResp(http.StatusForbidden, errors.New("permission denied"), "")
+	}
```
Comment (yuri-tceretian) [inline pkg/services/ngalert/api/api_alertmanager.go:131]: This is not necessary anymore as it is covered in [authorization code](https://github.com/grafana/grafana/blob/c1dbe7617ce98e581de4ab89aadd4639755079f5/pkg/services/ngalert/api/authorization.go#L33)

**Review by gotjosh — APPROVED** [review]
LGTM, nice catch!

Please see my nits and this also warrants a changelog entry, I believe this is a bugfix as we should never panic.

File: pkg/services/ngalert/api/api_alertmanager.go [hunk 3]
```diff
@@ -50,6 +51,12 @@ func (srv AlertmanagerSrv) RouteGetAMStatus(c *models.ReqContext) response.Respo
 }
 
 func (srv AlertmanagerSrv) RouteCreateSilence(c *models.ReqContext, postableSilence apimodels.PostableSilence) response.Response {
+	err := postableSilence.Validate(strfmt.Default)
+	if err != nil {
+		srv.log.Error("argument failed validation", "err", err)
+		return ErrResp(http.StatusBadRequest, err, "Silence Failed Validation: %v", postableSilence)
```
[the remark this answers is not in the document]
Comment (joeblubaugh) [reply pkg/services/ngalert/api/api_alertmanager.go]: Thanks, this is a lot better.

**Comment by joeblubaugh** [comment]
Closing and re-opening this PR to trigger the CLA check, since it's frozen.

**Comment by joeblubaugh** [comment]
👍 incoming
